internal/engine/dockerprune/docker_pruner.go

```diff
@@ -35,57 +39,86 @@ func NewDockerPruner(dCli docker.Client) *DockerPruner {
 	return &DockerPruner{dCli: dCli}
 }
 
-func (dp *DockerPruner) DisableForTesting() {
-	dp.disabledForTesting = true
+func (dp *DockerPruner) DisabledForTesting(disabled bool) {
+	dp.disabledForTesting = disabled
 }
 
 func (dp *DockerPruner) OnChange(ctx context.Context, st store.RStore) {
 	if dp.disabledForTesting {
 		return
 	}
 
-	if dp.started {
+	if err := dp.sufficientVersionError(); err != nil {
+		if !dp.insufficientVersionErrLogged {
+			logger.Get(ctx).Infof(
+				"[Docker prune] Docker API version too low for Tilt to run Docker Prune:\n\t%v", err,
+			)
+			dp.insufficientVersionErrLogged = true
+		}
 		return
 	}
 
 	state := st.RLockState()
-	defer st.RUnlockState()
+	settings := state.DockerPruneSettings
+	inProgBuild := state.CurrentlyBuilding
+	curBuildCount := state.CompletedBuildCount
+	hasDockerBuild := state.HasDockerBuild()
+	nextToBuild := buildcontrol.NextManifestNameToBuild(state)
+	st.RUnlockState()
+
+	if !settings.Enabled {
+		return
+	}
 
-	if state.DockerPruneSettings.Disabled {
+	// If user doesn't have at least one Docker build, they probably don't care about pruning
+	if !hasDockerBuild {
 		return
 	}
 
-	if err := dp.sufficientVersionError(); err != nil {
+	// Don't prune while we're building or about to build something, in case of weird side-effects.
+	if inProgBuild != "" || nextToBuild != "" {
 		return
 	}
 
-	// wait until state has been kinda initialized, and there's at least one Docker build
-	if !state.TiltStartTime.IsZero() && state.LastTiltfileError() == nil && state.HasDockerBuild() {
-		dp.started = true
-		go func() {
-			select {
-			case <-time.After(time.Minute * 2):
-				dp.Prune(ctx, defaultMaxAge) // report once pretty soon after startup...
-			case <-ctx.Done():
-				return
-			}
-
-			for {
-				select {
-				case <-time.After(dockerPruneInterval):
-					// and once every <interval> thereafter
-					dp.Prune(ctx, defaultMaxAge)
-				case <-ctx.Done():
-					return
-				}
-			}
-		}()
+	// Prune as soon after startup as we can (waiting until we've built SOMETHING)
+	if dp.lastPruneTime.IsZero() && curBuildCount > 0 {
+		dp.PruneAndRecordState(ctx, settings.MaxAge, curBuildCount)
+		return
+	}
+
+	// "Prune every X builds" takes precedence over "prune every Y hours"
+	if settings.NumBuilds != 0 {
+		buildsSince := curBuildCount - dp.lastPruneBuildCount
+		if buildsSince >= settings.NumBuilds {
+			dp.PruneAndRecordState(ctx, settings.MaxAge, curBuildCount)
+		}
+		return
+	}
+
+	interval := settings.Interval
+	if interval == 0 {
```

Comment:
i feel like the defaulting approach here is a big ad-hoc and likely to accidentally miss a place...maybe these should be methods of the settings object? Or there should be a NewDefaultSettings() constructor?

Reply:
that makes sense. most of the defaulting work will happen in the tiltfile settings i think, so i'll address this in that PR.